com.unity.netcode.gameobjects/Runtime/Messaging/DeferredMessageManager.cs

```diff
@@ -0,0 +1,150 @@
+        /// </summary>
+        public virtual unsafe void DeferMessage(IDeferredMessageManager.TriggerType trigger, ulong key, FastBufferReader reader, ref NetworkContext context)
+        {
+            if (!m_Triggers.ContainsKey(trigger))
+            {
+                m_Triggers[trigger] = new Dictionary<ulong, TriggerInfo>();
+            }
+
+            var triggers = m_Triggers[trigger];
+            if (!triggers.ContainsKey(key))
+            {
+                triggers[key] = new TriggerInfo
+                {
+                    Expiry = Time.realtimeSinceStartup + m_NetworkManager.NetworkConfig.SpawnTimeout,
+                    TriggerData = new NativeList<TriggerData>(Allocator.Persistent)
+                };
+            }
+
+            triggers[key].TriggerData.Add(new TriggerData
+            {
+                Reader = new FastBufferReader(reader.GetUnsafePtr(), Allocator.Persistent, reader.Length),
+                Header = context.Header,
+                Timestamp = context.Timestamp,
+                SenderId = context.SenderId,
+                SerializedHeaderSize = context.SerializedHeaderSize
+            });
+        }
+
+        /// <summary>
+        /// Cleans up any trigger that's existed for more than a second.
+        /// These triggers were probably for situations where a request was received after a despawn rather than before a spawn.
+        /// </summary>
+        public virtual unsafe void CleanupStaleTriggers()
+        {
+            foreach (var kvp in m_Triggers)
+            {
+                ulong* staleKeys = stackalloc ulong[kvp.Value.Count];
+                int index = 0;
+                foreach (var kvp2 in kvp.Value)
+                {
+                    if (kvp2.Value.Expiry < Time.realtimeSinceStartup)
+                    {
+                        staleKeys[index++] = kvp2.Key;
+                        PurgeTrigger(kvp.Key, kvp2.Key, kvp2.Value);
+                    }
+                }
+
+                for (var i = 0; i < index; ++i)
+                {
+                    kvp.Value.Remove(staleKeys[i]);
+                }
+            }
+        }
+
+        protected virtual void PurgeTrigger(IDeferredMessageManager.TriggerType triggerType, ulong key, TriggerInfo triggerInfo)
+        {
+            if (NetworkLog.CurrentLogLevel <= LogLevel.Normal)
+            {
+                NetworkLog.LogWarning($"Deferred messages were received for a trigger of type {triggerType} with key {key}, but that trigger was not received within within {m_NetworkManager.NetworkConfig.SpawnTimeout} second(s).");
+            }
+
+            foreach (var data in triggerInfo.TriggerData)
+            {
+                data.Reader.Dispose();
+            }
+
+            triggerInfo.TriggerData.Dispose();
+        }
+
+        public virtual void ProcessTriggers(IDeferredMessageManager.TriggerType trigger, ulong key)
+        {
+            if (m_Triggers.ContainsKey(trigger))
```

Comment:
I suggest using m_Triggers.TryGetValue, as there is no point in doing two lookups one after the other, and the logic can be a little simplified and cleaned up and made faster